README.md
# cdk8s-cli

### A command-line-interface for CDK for Kubernetes

**cdk8s** is a software development framework for defining Kubernetes
applications using rich object-oriented APIs. It allows developers to leverage
the full power of software in order to define abstract components called
"constructs" which compose Kubernetes resources or other constructs into
higher-level abstractions.

To install the CLI use one of the following methods:

* Homebrew: `brew install cdk8s`
* npm: `npm i -g cdk8s-cli`
* yarn: `yarn global add cdk8s-cli`

> **Note:** This repository is the command-line tool for cdk8s that is published as [`cdk8s-cli`](https://www.npmjs.com/package/cdk8s-cli). For more general information about cdk8s, please see [cdk8s.io](https://cdk8s.io), or visit the umbrella repository located at [cdk8s-team/cdk8s](https://github.com/cdk8s-team/cdk8s).

## Documentation

See [cdk8s.io](https://cdk8s.io).

## License

This project is distributed under the [Apache License, Version 2.0](./LICENSE).

This module is part of the [cdk8s project](https://github.com/awslabs/cdk8s).


CHANGELOG.md
## [0.0.0](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.100...v0.0.0) (2022-08-31)

### [2.0.100](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.99...v2.0.100) (2022-08-30)

### [2.0.99](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.98...v2.0.99) (2022-08-29)

### [2.0.98](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.97...v2.0.98) (2022-08-28)


### Bug Fixes

* typescript app template does not compile on Windows ([#421](https://github.com/cdk8s-team/cdk8s-cli/issues/421)) ([43d9db1](https://github.com/cdk8s-team/cdk8s-cli/commit/43d9db1fffe554215b1ff8fce38e6cfe669bef60)), closes [cdk8s-team/cdk8s-cli#362](https://github.com/cdk8s-team/cdk8s-cli/issues/362)

### [2.0.97](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.96...v2.0.97) (2022-08-28)

### [2.0.96](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.95...v2.0.96) (2022-08-27)

### [2.0.95](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.94...v2.0.95) (2022-08-26)

### [2.0.94](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.93...v2.0.94) (2022-08-25)

### [2.0.93](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.92...v2.0.93) (2022-08-24)

### [2.0.92](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.91...v2.0.92) (2022-08-23)

### [2.0.91](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.90...v2.0.91) (2022-08-22)

### [2.0.90](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.89...v2.0.90) (2022-08-21)

### [2.0.89](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.88...v2.0.89) (2022-08-20)

### [2.0.88](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.87...v2.0.88) (2022-08-19)

### [2.0.87](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.86...v2.0.87) (2022-08-18)

### [2.0.86](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.85...v2.0.86) (2022-08-17)

### [2.0.85](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.84...v2.0.85) (2022-08-16)

### [2.0.84](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.83...v2.0.84) (2022-08-15)

### [2.0.83](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.82...v2.0.83) (2022-08-14)

### [2.0.82](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.81...v2.0.82) (2022-08-13)

### [2.0.81](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.80...v2.0.81) (2022-08-12)

### [2.0.80](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.79...v2.0.80) (2022-08-11)

### [2.0.79](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.78...v2.0.79) (2022-08-10)

### [2.0.78](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.77...v2.0.78) (2022-08-09)

### [2.0.77](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.76...v2.0.77) (2022-08-08)

### [2.0.76](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.75...v2.0.76) (2022-08-07)

### [2.0.75](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.74...v2.0.75) (2022-08-06)

### [2.0.74](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.73...v2.0.74) (2022-08-05)

### [2.0.73](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.72...v2.0.73) (2022-08-04)

### [2.0.72](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.71...v2.0.72) (2022-08-03)

### [2.0.71](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.70...v2.0.71) (2022-08-02)

### [2.0.70](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.69...v2.0.70) (2022-08-01)

### [2.0.69](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.68...v2.0.69) (2022-07-31)

### [2.0.68](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.67...v2.0.68) (2022-07-30)

### [2.0.67](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.66...v2.0.67) (2022-07-29)

### [2.0.66](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.65...v2.0.66) (2022-07-28)

### [2.0.65](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.64...v2.0.65) (2022-07-27)


### Bug Fixes

* **import:** fail to import CRD without a schema ([#417](https://github.com/cdk8s-team/cdk8s-cli/issues/417)) ([3a3fce7](https://github.com/cdk8s-team/cdk8s-cli/commit/3a3fce7f4fd15a20f9e2289e320ba3092eba9688))

### [2.0.64](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.63...v2.0.64) (2022-07-27)

### [2.0.63](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.62...v2.0.63) (2022-07-26)

### [2.0.62](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.61...v2.0.62) (2022-07-25)

### [2.0.61](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.60...v2.0.61) (2022-07-24)

### [2.0.60](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.59...v2.0.60) (2022-07-23)

### [2.0.59](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.58...v2.0.59) (2022-07-22)

### [2.0.58](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.57...v2.0.58) (2022-07-21)

### [2.0.57](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.56...v2.0.57) (2022-07-21)


### Bug Fixes

* **import:** missing versions when importing multi-version crd ([#387](https://github.com/cdk8s-team/cdk8s-cli/issues/387)) ([74cb37e](https://github.com/cdk8s-team/cdk8s-cli/commit/74cb37e8d7b8a9260c9e3afe217dbb0040efad28))

### [2.0.56](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.55...v2.0.56) (2022-07-20)

### [2.0.55](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.54...v2.0.55) (2022-07-19)

### [2.0.54](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.53...v2.0.54) (2022-07-18)

### [2.0.53](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.52...v2.0.53) (2022-07-17)

### [2.0.52](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.51...v2.0.52) (2022-07-16)

### [2.0.51](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.50...v2.0.51) (2022-07-15)

### [2.0.50](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.49...v2.0.50) (2022-07-14)

### [2.0.49](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.48...v2.0.49) (2022-07-13)

### [2.0.48](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.47...v2.0.48) (2022-07-12)

### [2.0.47](https://github.com/cdk8s-team/cdk8s-cli/compare/v2.0.46...v2.0.47) (2022-07-11)


.projenrc.ts
import { Cdk8sCommon } from '@cdk8s/projen-common';
import { github, typescript, JsonFile, DependencyType } from 'projen';
import { addIntegTests } from './projenrc/integ';

const project = new typescript.TypeScriptProject({
  ...Cdk8sCommon.props,

  projenrcTs: true,
  name: 'cdk8s-cli',
  description: 'This is the command line tool for Cloud Development Kit (CDK) for Kubernetes (cdk8s).',
  repositoryUrl: 'https://github.com/cdk8s-team/cdk8s-cli.git',
  projenUpgradeSecret: 'PROJEN_GITHUB_TOKEN',
  authorName: 'Amazon Web Services',
  authorUrl: 'https://aws.amazon.com',
  minNodeVersion: '14.17.0',

  keywords: [
    'k8s',
    'cdk8s',
    'kubernetes',
    'cli',
    'tools',
    'automation',
    'containers',
  ],

  workflowBootstrapSteps: [{ run: 'pip3 install pipenv' }],

  defaultReleaseBranch: '2.x',
  majorVersion: 2,
  releaseBranches: {
    '1.x': {
      majorVersion: 1,
      npmDistTag: 'latest-1',
    },
  },

  releaseToNpm: true,
  bin: {
    cdk8s: 'bin/cdk8s',
  },
  deps: [
    'cdk8s',
    'cdk8s-plus-22',
    'codemaker',
    'constructs',
    'fs-extra@^8',
    'jsii-srcmak',
    'jsii-pacmak',
    'sscaff',
    'yaml',
    'yargs@^15',
    'json2jsii',
    'colors',
    'ajv',
    'table',
  ],
  devDeps: [
    '@cdk8s/projen-common',
    '@types/fs-extra@^8',
    '@types/json-schema',
    'glob',
    '@types/glob',
    'typescript-json-schema',
  ],

  tsconfig: {
    include: ['src/schemas/*.json'],
  },

  // run upgrade-dependencies workflow at a different hour than other cdk8s
  // repos to decrease flakiness of integration tests caused by new versions of
  // cdk8s and cdk8s+ being published to different languages at the same time
  depsUpgradeOptions: {
    // the latest versions of yaml require node > 12, which
    // is a change we are still not willing to make.
    exclude: ['yaml'],
    workflowOptions: {
      schedule: Cdk8sCommon.upgradeScheduleFor('cdk8s-cli'),
    },
  },
});

// ignore integration tests since they need to executed after packaging
// and are defined in a separate tasks.
project.jest?.addIgnorePattern('/test/integ/');

project.gitignore.exclude('.vscode/');

new Cdk8sCommon(project);

// add @types/node as a regular dependency since it's needed to during "import"
// to compile the generated jsii code.
project.deps.removeDependency('@types/node', DependencyType.BUILD);
project.deps.addDependency('@types/node@^12', DependencyType.RUNTIME);

const schemas = project.addTask('schemas');
schemas.exec('ts-node scripts/crd.schema.ts');

project.compileTask.spawn(schemas);

// run backport in clean directories every time.
const backportHome = '/tmp/.backport/';
const backportDir = `${backportHome}/repositories/cdk8s-team/cdk8s-cli`;
const backportConfig = new JsonFile(project, '.backportrc.json', {
  // see https://github.com/sqren/backport/blob/main/docs/config-file-options.md
  obj: {
    repoOwner: 'cdk8s-team',
    repoName: 'cdk8s-cli',
    signoff: true,
    branchLabelMapping: {
      '^backport-to-(.+)$': '$1',
    },
    prTitle: '{commitMessages}',
    fork: false,
    publishStatusCommentOnFailure: true,
    publishStatusCommentOnSuccess: true,
    publishStatusCommentOnAbort: true,
    targetPRLabels: [project.autoApprove!.label],
    dir: backportDir,
  },
});

// backport task to branches based on pr labels
const backportTask = createBackportTask();

// backport tasks to the explicit release branches
for (const branch of project.release!.branches) {
  createBackportTask(branch);
}

const backportWorkflow = project.github!.addWorkflow('backport');
backportWorkflow.on({ pullRequestTarget: { types: ['closed'] } });
backportWorkflow.addJob('backport', {
  runsOn: ['ubuntu-18.04'],
  permissions: {
    contents: github.workflows.JobPermission.WRITE,
  },
  steps: [
    // needed in order to run the projen task as well
    // as use the backport configuration in the repo.
    {
      name: 'checkout',
      uses: 'actions/checkout@v3',
      with: {
        // required because we need the full history
        // for proper backports.
        'fetch-depth': 0,
      },
    },
    {
      name: 'Set Git Identity',
      run: 'git config --global user.name "github-actions" && git config --global user.email "github-actions@github.com"',
    },
    {
      name: 'backport',
      if: 'github.event.pull_request.merged == true',
      run: `npx projen ${backportTask.name}`,
      env: {
        GITHUB_TOKEN: '${{ secrets.PROJEN_GITHUB_TOKEN }}',
        BACKPORT_PR_NUMBER: '${{ github.event.pull_request.number }}',
      },
    },
  ],
});

function createBackportTask(branch?: string) {
  const name = branch ? `backport:${branch}` : 'backport';
  const task = project.addTask(name, { requiredEnv: ['BACKPORT_PR_NUMBER', 'GITHUB_TOKEN'] });
  task.exec(`rm -rf ${backportHome}`);
  task.exec(`mkdir -p ${backportHome}`);
  task.exec(`cp ${backportConfig.path} ${backportHome}`);

  const command = ['npx', 'backport', '--accesstoken', '${GITHUB_TOKEN}', '--pr', '${BACKPORT_PR_NUMBER}'];
  if (branch) {
    command.push(...['--branch', branch]);
  } else {
    command.push('--non-interactive');
  }
  task.exec(command.join(' '), { cwd: backportHome });
  return task;
}

addIntegTests(project);

project.synth();
